### ISSUE TITLE
saving and loading params yields worse results

### ISSUE THREAD

**Issue by neiledgard** [apache/incubator-mxnet#6010]
## Environment info
Ubuntu 16.04.2 LTS \n \l
Linux as 4.10.0-041000-generic #201702191831 SMP Sun Feb 19 23:33:19 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.4)
Python 2.7.12

MXNet commit hash:
fdadaca2785068abcc2f31101769e7e065d86776

GPUs:
4 x GTX 1080 Ti

## Symptom:
Image classification example is running and saving after every epoch.
Eventually after stopping and loading an already saved epoch will result in a degrading model performance:

> cd /root/mxnet/example/image-classification
> python train_imagenet.py --data-train /mnt/fast/02_train.rec --data-val /mnt/fast/02_val.rec --gpu 0,1,2,3 --network resnet --batch-size 158 --lr 0.045 --disp-batches 5 --model an --model-prefix ./resnet01 --num-classes 5 --data-nthreads 12 --image-shape "3,256,256" --rgb-mean "159,199,191" --kv-store device --top-k 3 --num-epochs 1000 --load-epoch 28
> 

Output:

> INFO:root:start with arguments Namespace(batch_size=158, benchmark=0, data_nthreads=12, data_train='/mnt/fast/02_train.rec', data_val='/mnt/fast/02_val.rec', disp_batches=5, dtype='float32', gpus='0,1,2,3', image_shape='3,256,256', kv_store='device', load_epoch=28, lr=0.045, lr_factor=0.1, lr_step_epochs='30,60', max_random_aspect_ratio=0.25, max_random_h=36, max_random_l=50, max_random_rotate_angle=10, max_random_s=50, max_random_scale=1, max_random_shear_ratio=0.1, min_random_scale=1, model_prefix='./resnet01', mom=0.9, monitor=0, network='resnet', num_classes=5, num_epochs=1000, num_examples=1281167, num_layers=50, optimizer='sgd', pad_size=0, random_crop=1, random_mirror=1, rgb_mean='159,199,191', test_io=0, top_k=3, wd=0.0001)
> [19:27:43] src/io/iter_image_recordio_2.cc:135: ImageRecordIOParser2: /mnt/fast/02_train.rec, use 5 threads for decoding..
> [19:27:44] src/io/iter_image_recordio_2.cc:135: ImageRecordIOParser2: /mnt/fast/02_val.rec, use 5 threads for decoding..
> INFO:root:Loaded model ./resnet01_0028.params
> 
> 
> [19:28:48] src/operator/././cudnn_algoreg-inl.h:65: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
> INFO:root:Epoch[28] Batch [5]   Speed: 610.95 samples/sec       Train-accuracy=0.978903
> INFO:root:Epoch[28] Batch [5]   Speed: 610.95 samples/sec       Train-top_k_accuracy_3=0.998945
> INFO:root:Epoch[28] Batch [10]  Speed: 534.55 samples/sec       Train-accuracy=0.955696
> INFO:root:Epoch[28] Batch [10]  Speed: 534.55 samples/sec       Train-top_k_accuracy_3=0.997468
> INFO:root:Epoch[28] Batch [15]  Speed: 536.39 samples/sec       Train-accuracy=0.925316
> INFO:root:Epoch[28] Batch [15]  Speed: 536.39 samples/sec       Train-top_k_accuracy_3=0.991139
> INFO:root:Epoch[28] Batch [20]  Speed: 527.45 samples/sec       Train-accuracy=0.859494
> INFO:root:Epoch[28] Batch [20]  Speed: 527.45 samples/sec       Train-top_k_accuracy_3=0.979747
> INFO:root:Epoch[28] Batch [25]  Speed: 545.23 samples/sec       Train-accuracy=0.806329
> INFO:root:Epoch[28] Batch [25]  Speed: 545.23 samples/sec       Train-top_k_accuracy_3=0.958228
> INFO:root:Epoch[28] Batch [30]  Speed: 534.52 samples/sec       Train-accuracy=0.729114
> INFO:root:Epoch[28] Batch [30]  Speed: 534.52 samples/sec       Train-top_k_accuracy_3=0.949367
> INFO:root:Epoch[28] Batch [35]  Speed: 529.25 samples/sec       Train-accuracy=0.654430
> INFO:root:Epoch[28] Batch [35]  Speed: 529.25 samples/sec       Train-top_k_accuracy_3=0.927848
> mxnet/example/image-classification/resnet01-0003.params
> INFO:root:Epoch[28] Batch [90]  Speed: 525.33 samples/sec       Train-top_k_accuracy_3=0.937975
> INFO:root:Epoch[28] Batch [95]  Speed: 530.39 samples/sec       Train-accuracy=0.758228
> INFO:root:Epoch[28] Batch [95]  Speed: 530.39 samples/sec       Train-top_k_accuracy_3=0.946835
> INFO:root:Epoch[28] Batch [100] Speed: 533.11 samples/sec       Train-accuracy=0.731646
> INFO:root:Epoch[28] Batch [100] Speed: 533.11 samples/sec       Train-top_k_accuracy_3=0.951899
> INFO:root:Epoch[28] Batch [105] Speed: 520.15 samples/sec       Train-accuracy=0.755696
> INFO:root:Epoch[28] Batch [105] Speed: 520.15 samples/sec       Train-top_k_accuracy_3=0.951899
> INFO:root:Epoch[28] Batch [110] Speed: 541.82 samples/sec       Train-accuracy=0.800000
> INFO:root:Epoch[28] Batch [110] Speed: 541.82 samples/sec       Train-top_k_accuracy_3=0.968354
> INFO:root:Epoch[28] Batch [115] Speed: 530.67 samples/sec       Train-accuracy=0.788608
> INFO:root:Epoch[28] Batch [115] Speed: 530.67 samples/sec       Train-top_k_accuracy_3=0.965823
> INFO:root:Epoch[28] Batch [120] Speed: 520.34 samples/sec       Train-accuracy=0.759494

It produces similar behavior when using some other networks as well. I suppose it is because of batch normalization but that also doesn't really make sense.

If you have any suggestions what I might doing wrong, please let me know.

**Comment by piiswrong** [comment]
can you describe the log? What is show degraded perf?

**Comment by neiledgard** [comment]
Before exiting the previous training (that ended at epoch 28), the training accuracy increases up to around 0.97. After exiting and reloading epoch 28, you can see training accuracy going down

> 
> Train-accuracy=0.978903
> Train-accuracy=0.925316
> ...
> Train-accuracy=0.731646
> Train-accuracy=0.654430
> 

By keeping it run, it will converge back to around 0.97:

> INFO:root:Epoch[28] Batch [18090]	Speed: 518.36 samples/sec	Train-accuracy=0.977215
> INFO:root:Epoch[28] Batch [18090]	Speed: 518.36 samples/sec	Train-top_k_accuracy_3=0.997468

**Comment by piiswrong** [comment]
Most like because you didn't decrease learning rate

**Comment by neiledgard** [comment]
As of my best understanding, it shouldn't matter, from epoch 0 to 30 lr doesn't change either way. Also it doesn't matter where I cancel the training and reload the last epoch's parameters, it is always much lower perf compared to if I just don't cancel it.

**Comment by piiswrong** [comment]
@mli Get someone to take a look?

**Comment by neiledgard** [comment]
Is it possible that its the batch normalizing layer? It is suspicious, especially because I use 4 GPUs.

**Comment by neiledgard** [comment]
same with resnet

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
